File: Iverson.Clients/DotNet/Iverson.Client.Search/GroupByBuilder.cs
using Iverson.Client.Contracts;

namespace Iverson.Client.Search;

/// <summary>
/// Fluent DSL builder that compiles to a <see cref="GroupByRequest"/>.
/// Not generic on a type parameter — joins bring multiple registered types into scope,
/// so keys, filters, and metrics are addressed by raw field-name strings.
/// </summary>
public sealed class GroupByBuilder
{
    private readonly string             _typeName;
    private readonly List<string>       _keys    = [];
    private readonly List<MetricSpec>   _metrics = [];
    private readonly List<SearchSort>   _orderBy = [];
    private readonly List<SearchClause> _where   = [];
    private readonly List<SearchClause> _having  = [];
    private readonly List<JoinSpec>     _joins   = [];
    private SearchLogic _whereLogic  = SearchLogic.And;
    private SearchLogic _havingLogic = SearchLogic.And;
    private int          _limit      = 10_000;

    public GroupByBuilder(string typeName) => _typeName = typeName;

    // ── Keys ────────────────────────────────────────────────────────────────────

    /// <summary>Adds a single GROUP BY key.</summary>
    public GroupByBuilder Key(string field)
    {
        _keys.Add(field);
        return this;
    }

    /// <summary>Adds multiple GROUP BY keys.</summary>
    public GroupByBuilder Keys(params string[] fields)
    {
        _keys.AddRange(fields);
        return this;
    }

    // ── WHERE filter (raw field strings, same operators as QueryBuilder) ─────────

    public GroupByBuilder Where(string field, SearchOperator op, object value)
        => AddWhere(field, op, value, SearchClauseType.Filter);

    /// <summary>Adds a MUST_NOT WHERE clause (excludes matches before grouping).</summary>
    public GroupByBuilder Not(string field, SearchOperator op, object value)
        => AddWhere(field, op, value, SearchClauseType.MustNot);

    public GroupByBuilder WithLogic(SearchLogic logic)
    {
        _whereLogic = logic;
        return this;
    }

    // ── HAVING (references output alias names) ────────────────────────────────────

    public GroupByBuilder Having(string alias, SearchOperator op, object value)
        => AddHaving(alias, op, value);

    /// <summary>Sets the logic combining HAVING clauses. Defaults to AND.</summary>
    public GroupByBuilder WithHavingLogic(SearchLogic logic)
    {
        _havingLogic = logic;
        return this;
    }

    // ── JOIN ────────────────────────────────────────────────────────────────────

    /// <summary>
    /// Adds a join from this builder's own type (<see cref="_typeName"/>) to
    /// <paramref name="rightType"/> via the given fields. Matches the
    /// leftField/rightType/rightField ordering used by the Java, Python, TypeScript, and
    /// Go clients — leftType is inferred rather than passed explicitly, which intentionally
    /// drops the ability to express a join from a non-primary already-joined table (multi-hop)
    /// in this builder, matching those clients' scope.
    /// </summary>
    public GroupByBuilder Join(
        string leftField, string rightType, string rightField,
        JoinKind kind = JoinKind.Inner)
    {
        _joins.Add(new JoinSpec
        {
            LeftType   = _typeName,
            RightType  = rightType,
            LeftField  = leftField,
            RightField = rightField,
            Kind       = kind
        });
        return this;
    }

    // ── Metrics — simple field ─────────────────────────────────────────────────

    public GroupByBuilder Sum(string field, string? alias = null)
        => AddMetric(alias ?? $"{field}_sum", AggregationType.Sum, field: field);

    public GroupByBuilder Avg(string field, string? alias = null)
        => AddMetric(alias ?? $"{field}_avg", AggregationType.Avg, field: field);

    public GroupByBuilder Min(string field, string? alias = null)
        => AddMetric(alias ?? $"{field}_min", AggregationType.Min, field: field);

    public GroupByBuilder Max(string field, string? alias = null)
        => AddMetric(alias ?? $"{field}_max", AggregationType.Max, field: field);

    public GroupByBuilder Count(string field, string? alias = null)
        => AddMetric(alias ?? $"{field}_count", AggregationType.Count, field: field);

    /// <summary>COUNT(*) — leaves the metric's field empty.</summary>
    public GroupByBuilder CountAll(string alias = "count")
        => AddMetric(alias, AggregationType.Count);

    // ── Metrics — expression (raw SQL) ─────────────────────────────────────────

    public GroupByBuilder SumExpr(string expression, string alias)
        => AddMetric(alias, AggregationType.Sum, expression: expression);

    public GroupByBuilder AvgExpr(string expression, string alias)
        => AddMetric(alias, AggregationType.Avg, expression: expression);

    // ── Sorting and limit ───────────────────────────────────────────────────────

    public GroupByBuilder OrderBy(string field, bool descending = false)
    {
        _orderBy.Add(new SearchSort { Property = field, Descending = descending });
        return this;
    }

    public GroupByBuilder Limit(int n)
    {
        _limit = n;
        return this;
    }

    // ── Build ──────────────────────────────────────────────────────────────────

    public GroupByRequest Build(string? traceId = null)
    {
        var aliasSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var m in _metrics)
            if (!aliasSet.Add(m.Name))
                throw new InvalidOperationException($"Duplicate metric alias '{m.Name}'.");

        var keySet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var k in _keys)
        {
            if (!keySet.Add(k))
                throw new InvalidOperationException($"Duplicate key '{k}'.");
            if (!aliasSet.Add(k))
                throw new InvalidOperationException($"Key '{k}' collides with an existing metric alias.");
        }

        foreach (var h in _having)
            if (!aliasSet.Contains(h.Property))
                throw new InvalidOperationException(
                    $"HAVING references '{h.Property}', which is neither a metric alias nor a key.");
        foreach (var s in _orderBy)
            if (!aliasSet.Contains(s.Property))
                throw new InvalidOperationException(
                    $"OrderBy references '{s.Property}', which is neither a metric alias nor a key.");

        var request = new GroupByRequest
        {
            TypeName = _typeName,
            Query    = new SearchQuery { Logic = _whereLogic },
            Having   = new SearchQuery { Logic = _havingLogic },
            Limit    = _limit,
            TraceId  = traceId ?? string.Empty
        };
        request.Keys.AddRange(_keys);
        request.Metrics.AddRange(_metrics);
        request.Query.Clauses.AddRange(_where);
        request.Having.Clauses.AddRange(_having);
        request.OrderBy.AddRange(_orderBy);
        request.Joins.AddRange(_joins);
        return request;
    }

    // ── Private helpers ────────────────────────────────────────────────────────

    private GroupByBuilder AddMetric(
        string alias, AggregationType type, string? field = null, string? expression = null)
    {
        var spec = new MetricSpec { Name = alias, Type = type };
        if (field is not null)      spec.Field      = field;
        if (expression is not null) spec.Expression = expression;
        _metrics.Add(spec);
        return this;
    }

    private GroupByBuilder AddWhere(string field, SearchOperator op, object value, SearchClauseType clauseType)
    {
        _where.Add(new SearchClause
        {
            Property   = field,
            Operator   = op,
            Value      = SearchValueConverter.ToSearchValue(value),
            ClauseType = clauseType
        });
        return this;
    }

    private GroupByBuilder AddHaving(string alias, SearchOperator op, object value)
    {
        _having.Add(new SearchClause
        {
            Property   = alias,
            Operator   = op,
            Value      = SearchValueConverter.ToSearchValue(value),
            ClauseType = SearchClauseType.Filter
        });
        return this;
    }
}

File: Iverson.Clients/Java/client/src/main/java/io/iverson/client/search/GroupByBuilder.java
package io.iverson.client.search;

import iverson.ObjectSearch.AggregationType;
import iverson.ObjectSearch.JoinKind;
import iverson.ObjectSearch.JoinSpec;
import iverson.ObjectSearch.MetricSpec;
import iverson.ObjectSearch.SearchClause;
import iverson.ObjectSearch.SearchClauseType;
import iverson.ObjectSearch.SearchLogic;
import iverson.ObjectSearch.SearchOperator;
import iverson.ObjectSearch.SearchQuery;
import iverson.ObjectSearch.SearchSort;
import iverson.ObjectSearch.GroupByRequest;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Fluent DSL builder that compiles to a {@link GroupByRequest} proto message.
 *
 * <p>Unlike {@link QueryBuilder}, this builder is not generic on a type parameter —
 * joins bring multiple registered types into scope, so keys, filters, and metrics are
 * addressed by raw field-name strings.</p>
 *
 * <p>Does not require a live server — {@link #build()} simply returns the compiled proto.
 * Instantiate via {@link Query#groupBy(String)}.</p>
 *
 * <pre>{@code
 * GroupByRequest req = Query.groupBy("LineItem")
 *     .keys("returnFlag", "lineStatus")
 *     .sum("quantity")
 *     .avg("price")
 *     .countAll()
 *     .orderBy("returnFlag")
 *     .build();
 * }</pre>
 */
public final class GroupByBuilder {

    private final String typeName;
    private final List<String>       keys     = new ArrayList<>();
    private final List<MetricSpec>   metrics  = new ArrayList<>();
    private final List<SearchSort>   orderBy  = new ArrayList<>();
    private final List<SearchClause> where    = new ArrayList<>();
    private final List<SearchClause> having   = new ArrayList<>();
    private final List<JoinSpec>     joins    = new ArrayList<>();
    private SearchLogic whereLogic  = SearchLogic.AND;
    private SearchLogic havingLogic = SearchLogic.AND;
    private int         limit       = 10_000;

    GroupByBuilder(String typeName) {
        this.typeName = typeName;
    }

    // ── Keys ──────────────────────────────────────────────────────────────────

    /** Adds a single GROUP BY key. */
    public GroupByBuilder key(String field) {
        keys.add(field);
        return this;
    }

    /** Adds multiple GROUP BY keys. */
    public GroupByBuilder keys(String... fields) {
        keys.addAll(Arrays.asList(fields));
        return this;
    }

    // ── WHERE filter (raw field strings, same operators/encoding as QueryBuilder) ─

    /** Adds a WHERE (FILTER) clause. Uses the same value encoding as {@link QueryBuilder}. */
    public GroupByBuilder where(String field, SearchOperator op, Object value) {
        where.add(SearchClause.newBuilder()
            .setProperty(field)
            .setOperator(op)
            .setValue(SearchValues.toSearchValue(value))
            .setClauseType(SearchClauseType.FILTER)
            .build());
        return this;
    }

    /** Adds a MUST_NOT WHERE clause (excludes matches before grouping). */
    public GroupByBuilder not(String field, SearchOperator op, Object value) {
        where.add(SearchClause.newBuilder()
            .setProperty(field)
            .setOperator(op)
            .setValue(SearchValues.toSearchValue(value))
            .setClauseType(SearchClauseType.MUST_NOT)
            .build());
        return this;
    }

    /** Sets the logic used to combine top-level WHERE clauses. Defaults to AND. */
    public GroupByBuilder withLogic(SearchLogic logic) {
        this.whereLogic = logic;
        return this;
    }

    // ── HAVING (references output alias names) ──────────────────────────────────

    /** Adds a HAVING clause. {@code alias} must match a metric's output alias. */
    public GroupByBuilder having(String alias, SearchOperator op, Object value) {
        having.add(SearchClause.newBuilder()
            .setProperty(alias)
            .setOperator(op)
            .setValue(SearchValues.toSearchValue(value))
            .setClauseType(SearchClauseType.FILTER)
            .build());
        return this;
    }

    /** Sets the logic combining HAVING clauses. Defaults to AND. */
    public GroupByBuilder withHavingLogic(SearchLogic logic) {
        this.havingLogic = logic;
        return this;
    }

    // ── JOIN ──────────────────────────────────────────────────────────────────

    /** Adds an INNER join from this type to {@code rightType} on the given fields. */
    public GroupByBuilder join(String leftField, String rightType, String rightField) {
        return join(leftField, rightType, rightField, JoinKind.INNER);
    }

    /** Adds a join of the given {@link JoinKind} from this type to {@code rightType}. */
    public GroupByBuilder join(String leftField, String rightType, String rightField, JoinKind kind) {
        joins.add(JoinSpec.newBuilder()
            .setLeftType(typeName)
            .setRightType(rightType)
            .setLeftField(leftField)
            .setRightField(rightField)
            .setKind(kind)
            .build());
        return this;
    }

    // ── Metrics — simple field ───────────────────────────────────────────────────

    public GroupByBuilder sum(String field) {
        return sum(field, field + "_sum");
    }

    public GroupByBuilder sum(String field, String alias) {
        return addMetric(alias, AggregationType.SUM, field, null);
    }

    public GroupByBuilder avg(String field) {
        return avg(field, field + "_avg");
    }

    public GroupByBuilder avg(String field, String alias) {
        return addMetric(alias, AggregationType.AVG, field, null);
    }

    public GroupByBuilder min(String field) {
        return min(field, field + "_min");
    }

    public GroupByBuilder min(String field, String alias) {
        return addMetric(alias, AggregationType.MIN, field, null);
    }

    public GroupByBuilder max(String field) {
        return max(field, field + "_max");
    }

    public GroupByBuilder max(String field, String alias) {
        return addMetric(alias, AggregationType.MAX, field, null);
    }

    public GroupByBuilder count(String field) {
        return count(field, field + "_count");
    }

    public GroupByBuilder count(String field, String alias) {
        return addMetric(alias, AggregationType.COUNT, field, null);
    }

    /** COUNT(*) — leaves the metric's field empty. Alias defaults to {@code "count"}. */
    public GroupByBuilder countAll() {
        return countAll("count");
    }

    /** COUNT(*) — leaves the metric's field empty. */
    public GroupByBuilder countAll(String alias) {
        return addMetric(alias, AggregationType.COUNT, null, null);
    }

    // ── Metrics — expression (raw SQL) ───────────────────────────────────────────

    /** SUM over a raw SQL expression, e.g. {@code "price * (1 - discount)"}. */
    public GroupByBuilder sumExpr(String expression, String alias) {
        return addMetric(alias, AggregationType.SUM, null, expression);
    }

    /** AVG over a raw SQL expression. */
    public GroupByBuilder avgExpr(String expression, String alias) {
        return addMetric(alias, AggregationType.AVG, null, expression);
    }

    // ── Sorting and limit ─────────────────────────────────────────────────────────

    public GroupByBuilder orderBy(String field) {
        return orderBy(field, false);
    }

    public GroupByBuilder orderByDesc(String field) {
        return orderBy(field, true);
    }

    public GroupByBuilder orderBy(String field, boolean desc) {
        orderBy.add(SearchSort.newBuilder().setProperty(field).setDescending(desc).build());
        return this;
    }

    public GroupByBuilder limit(int n) {
        this.limit = n;
        return this;
    }

    // ── Build ─────────────────────────────────────────────────────────────────────

    /** Builds and returns the {@link GroupByRequest} proto message. */
    public GroupByRequest build() {
        return build("");
    }

    /** Builds and returns the {@link GroupByRequest} proto message with the given trace ID. */
    public GroupByRequest build(String traceId) {
        java.util.Set<String> aliases = new java.util.HashSet<>();
        for (MetricSpec m : metrics)
            if (!aliases.add(m.getName().toLowerCase(java.util.Locale.ROOT)))
                throw new IllegalStateException("Duplicate metric alias '" + m.getName() + "'.");
        for (String k : keys) aliases.add(k.toLowerCase(java.util.Locale.ROOT));

        for (SearchClause h : having)
            if (!aliases.contains(h.getProperty().toLowerCase(java.util.Locale.ROOT)))
                throw new IllegalStateException("HAVING references '" + h.getProperty()
                    + "', which is neither a metric alias nor a key.");
        for (SearchSort s : orderBy)
            if (!aliases.contains(s.getProperty().toLowerCase(java.util.Locale.ROOT)))
                throw new IllegalStateException("orderBy references '" + s.getProperty()
                    + "', which is neither a metric alias nor a key.");

        SearchQuery query = SearchQuery.newBuilder()
            .addAllClauses(where)
            .setLogic(whereLogic)
            .build();

        SearchQuery havingQuery = SearchQuery.newBuilder()
            .addAllClauses(having)
            .setLogic(havingLogic)
            .build();

        return GroupByRequest.newBuilder()
            .setTypeName(typeName)
            .setQuery(query)
            .addAllKeys(keys)
            .addAllMetrics(metrics)
            .setHaving(havingQuery)
            .addAllOrderBy(orderBy)
            .setLimit(limit)
            .addAllJoins(joins)
            .setTraceId(traceId == null ? "" : traceId)
            .build();
    }

    // ── Internal helpers ─────────────────────────────────────────────────────────

    private GroupByBuilder addMetric(String alias, AggregationType type, String field, String expression) {
        MetricSpec.Builder spec = MetricSpec.newBuilder()
            .setName(alias)
            .setType(type);
        if (field != null) spec.setField(field);
        if (expression != null) spec.setExpression(expression);
        metrics.add(spec.build());
        return this;
    }
}
